osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/firmware/UpdateFirmwareCommandExecutor.java

```diff
@@ -118,20 +124,98 @@ private void activate(final ImageTransfer transfer) throws OsgpException {
     }
   }
 
-  private byte[] getImageData(final String firmwareIdentification) throws ProtocolAdapterException {
-    final byte[] firmwareFile = this.firmwareFileCachingRepository.retrieve(firmwareIdentification);
+  private FirmwareFile getFirmwareFile(
+      final UpdateFirmwareRequestDto updateFirmwareRequestDto,
+      final MessageMetadata messageMetadata)
+      throws ProtocolAdapterException {
+
+    log.debug(
+        "Getting firmware file from caching repository for firmware {}",
+        updateFirmwareRequestDto.getFirmwareIdentification());
+    final byte[] firmwareFileByteArray =
+        this.firmwareFileCachingRepository.retrieve(
+            updateFirmwareRequestDto.getFirmwareIdentification());
 
-    if (firmwareFile == null) {
+    if (firmwareFileByteArray == null) {
       throw new ProtocolAdapterException(EXCEPTION_MSG_FIRMWARE_FILE_NOT_AVAILABLE);
     }
 
+    final FirmwareFile firmwareFile = new FirmwareFile(firmwareFileByteArray);
+    if (firmwareFile.isMbusFirmware()) {
+      this.addMac(
+          messageMetadata, updateFirmwareRequestDto.getDeviceIdentification(), firmwareFile);
+    }
+
+    return firmwareFile;
+  }
+
+  private FirmwareFile addMac(
+      final MessageMetadata messageMetadata,
+      final String deviceIdentification,
+      final FirmwareFile firmwareFile)
+      throws ProtocolAdapterException {
+
+    log.debug("Adding MAC to firmware file for M-Bus device");
+    final DlmsDevice mbusDevice =
+        this.dlmsDeviceRepository.findByDeviceIdentification(deviceIdentification);
+
+    final Long mbusIdentificationNumber =
+        Long.parseLong(String.valueOf(mbusDevice.getMbusIdentificationNumber()), 16);
+    log.debug(
+        "Setting M-Bus device Identification number: {} (hex:{})",
+        mbusIdentificationNumber.intValue(),
+        mbusDevice.getMbusIdentificationNumber());
+    firmwareFile.setMbusDeviceIdentificationNumber(mbusIdentificationNumber.intValue());
+
+    final byte[] calculatedMac =
+        this.macGenerationService.calculateMac(
+            messageMetadata, mbusDevice.getDeviceIdentification(), firmwareFile);
+
+    log.debug("Calculated MAC: {}", Arrays.toString(calculatedMac));
+
+    firmwareFile.setSecurityByteArray(calculatedMac);
+
     return firmwareFile;
   }
 
+  private String getImageIdentifier(
+      final String firmwareIdentification, final FirmwareFile firmwareFile)
+      throws ProtocolAdapterException {
+
+    String imageIdentifier = null;
+
+    if (firmwareFile.isMbusFirmware()) {
+
+      imageIdentifier = firmwareFile.createImageIdentifierForMbusDevice();
```

Comment:
This will be refactored;
imageIdentifier returned here will be a byte array (see earlier comment) and printed as a array of bytes `Arrays.toString(imageIdentifier)`

Reply:
Has been refactored.